### ISSUE TITLE
Applying 2.1.1 Keyboard on Mobile

### ISSUE THREAD

**Issue by Moaan** [w3c/wcag2ict#288]
Hello,
I am doing research on mobile keyboard testing and I’d like to better understand how WCAG success criterion 2.1.1 Keyboard should be applied to iOS and Android operating systems and how to test for it. 

The intent mentions “content can be operated through a keyboard or keyboard interface (so an alternate keyboard can be used).” 

Before iOS 13.4 was released--when Apple added ‘Full Keyboard Access’--what was the keyboard interface? Did VoiceOver screen reader testing count as testing against the keyboard interface because users were able to operate VoiceOver with a connected keyboard?

If that is the case, is it technically sufficient today if functionality is operable with VoiceOver but not with Full Keyboard Access ?

**Comment by bruce-usab** [comment]
@Moaan on-screen keyboards were available in 2008 when WCAG 2.0 and SC 2.1.1 was finalized.  2.1.1 has not changed with WCAG 2.1 or 2.2.  The 2013 WCAG2ICT provides some analysis but [applies directly as written](https://www.w3.org/TR/wcag2ict/#keyboard-operation-keyboard-operable).

iOS and Android have long facilitated the option to use an external keyboard, so that's probably not the barrier.  The new "Full Keyboard Access" is more functionally, but I do not think it is determinative of "functionality of the content is operable through a keyboard interface".

If an iOS app is functionality operable with VoiceOver, there is probably no failure of 2.1.1.  But I am not following how the iOS Full Keyboard Access setting impacts that.

**Comment by GreggVan** [comment]
IBID for Bruce's comment. 
- external keyboard support has been available for many years on iOS and Android. 
-  mobile apps  RARELY provide their own keyboard -  and instead use the keyboard interface of the OS.   there are multiple different keyboards that will feed keystrokes to that interface - so that the APP will get them
- - the built-in keyboard on screen in iOS/Android
- - any of the installable on-screen keyboards you can install in place of the  iOS/Android keyboard  (this include AT keyboard replacements that take morse code or many other types of AT-like input and output keystrokes)
- - external keyboards you plug into your mobile device
- - eternal keyboards you connect via bluetooth 

does that help?

**Comment by patrickhlauke** [comment]
> If an iOS app is functionality operable with VoiceOver, there is probably no failure of 2.1.1

whether an app is VO operable is orthogonal to whether or not it's keyboard operable. those are separate aspects.

to determine if an iOS application is keyboard operable, you'll need to test it with an external/physical keyboard or keyboard-like interface. real-world keyboard users on iOS will have the full keyboard access setting enabled, so i'd say that this would be a reasonable starting point for determining 2.1.1 pass/fail

**Comment by mraccess77** [comment]
A keyboard interface allows users to send keystroke input to content regardless of whether there is a physical or on-screen keyboard.  There is an interfaces in iOS and several input methods that work with some aspects of that interface.

Nearly all on-screen keyboards I have found with iOS don't allow all manner keystrokes such as arrow keys to be sent to the iOS keyboard interface.  In addition, there are many places in the platform and Safari and apps where you can't launch an onscreen keyboard in order to send keystrokes to the interface as the on-screen keyboard display is restricted.

Prior to full keyboard access VoiceOver when active provided a way to interact with the keyboard interface via an external keyboard or braille display input.  iOS at the time implemented sticky keys and audio off to help users who wanted to use this method and did not need the speech interface. Braille users could also use a braille display input without speech in these cases.

iOS then introduced full keyboard support.  The fact that they introduced this mode tells me that there was a user need for such a mode.  This mode is one of the keyboard interface access methods that can be used with different forms of input - an external keyboard.

Typically interpretation of SC 2.1.1 has now allowed methods of controlling the pointer via the keyboard interface such as mouseKeys to be sufficient to qualify as a keyboard interface -because it is using a keyboard to control the pointer interface.  The gestures command in full keyboard access would seem to not meet this as it seems to be controlling the pointer by sending a gesture.

It was in the past possible to validate keyboard interface support with VoiceOver as long as VoiceOver actually controlling the keyboard interface and not the pointer interface.  This could still be argued today - however, given that the full keyboard feature in iOS was created and is available and is designed to support navigation without moving to static content - I would think it is preferred to validate functionality is working through this full keyboard feature - in particular with custom web user interface elements and non-web content.

It may be helpful to discuss the places where VoiceOver keyboard access works but full keyboard support fails.  For example, when you activate an element with VoiceOver with VO+space it may be sending a touch event rather than a keyboard event to the element.  If this were the case then the element would seem to not support the keyboard interface but only working with VoiceOver because VoiceOver is sending a pointer event.  This is the important aspect to understand in my opinion as the external keyboard works with VoiceOver but is it really using the keyboard interface?

**Comment by Moaan** [comment]
> **bruce-usab:** “If an iOS app is functionality operable with VoiceOver, there is probably no failure of 2.1.1. But I am not following how the iOS Full Keyboard Access setting impacts that.”

The implication is that a product can have a conformance report where 2.1.1 Keyboard is Supports even though it only works well with VoiceOver(VO) and not with Full Keyboard Access (FKA), even though FKA is the assistive technology made for mobile keyboard users. 
VO tends to work a lot better than FKA so unless mobile keyboard users are aware of this and have decided to just use VO with speech off, then this interpretation of 2.1.1 does not speak to the requirements of actual keyboard users. Research on this would be good to better understand mobile users who with motor impairments.

**Comment by Moaan** [comment]
> **mraccess77:** “Typically interpretation of SC 2.1.1 has not allowed methods of controlling the pointer via the keyboard interface such as mouseKeys to be sufficient to qualify as a keyboard interface -because it is using a keyboard to control the pointer interface. The gestures command in full keyboard access would seem to not meet this as it seems to be controlling the pointer by sending a gesture.”

Interesting, "Keyboard Gestures" mode is sometimes the only option. I guess if being strict about 2.1.1 this would be a failure but it can be claimed as equivalent facilitation for Functional Performance Criteria ? In practice, I would prefer to be lenient and just accept the use of keyboard gestures as normal keyboard usage even if it is not using the keyboard interface. 

> **mraccess77:** “It was in the past possible to validate keyboard interface support with VoiceOver as long as VoiceOver actually controlling the keyboard interface and not the pointer interface. This could still be argued today - however, given that the full keyboard feature in iOS was created and is available and is designed to support navigation without moving to static content - I would think it is preferred to validate functionality is working through this full keyboard feature - in particular with custom web user interface elements and non-web content.”

This answers my question. My takeaway is that even though VO may be technically enough, for the reasons you outlined, it is probably best to consider FKA support a requirement of 2.1.1 for an iOS app.

Thank you all for the thoughtful responses!

**Comment by bruce-usab** [comment]
@Moaan please close this issue since you write that your question has been answered.
